core/trino-main/src/main/java/io/trino/server/InternalAuthenticationManager.java

```diff
@@ -90,93 +81,72 @@ public InternalAuthenticationManager(String sharedSecret, String nodeId)
     {
         requireNonNull(sharedSecret, "sharedSecret is null");
         requireNonNull(nodeId, "nodeId is null");
-        this.hmac = hmacShaKeyFor(Hashing.sha256().hashString(sharedSecret, UTF_8).asBytes());
+        this.hashing = Hashing.hmacSha256(sharedSecret.getBytes(UTF_8));
         this.nodeId = nodeId;
-        this.jwtParser = newJwtParserBuilder().verifyWith(hmac).build();
-        this.currentToken = new AtomicReference<>(createJwt());
     }
 
     public static boolean isInternalRequest(ContainerRequestContext request)
     {
-        return request.getHeaders().getFirst(TRINO_INTERNAL_BEARER) != null;
+        return request.getHeaders().getFirst(TRINO_INTERNAL_SIGNATURE) != null;
     }
 
     public void handleInternalRequest(ContainerRequestContext request)
     {
-        String subject;
-        try {
-            subject = parseJwt(request.getHeaders().getFirst(TRINO_INTERNAL_BEARER));
-        }
-        catch (JwtException e) {
-            log.error(e, "Internal authentication failed");
+        String nodeId = getRequiredHeader(request, TRINO_INTERNAL_NODE_ID);
+        String signature = getRequiredHeader(request, TRINO_INTERNAL_SIGNATURE);
+
+        long requestTimestampMillis = parseLong(getRequiredHeader(request, TRINO_INTERNAL_REQUEST_TIMESTAMP));
+
+        if (!signature.equals(signature(nodeId, requestTimestampMillis, request.getUriInfo().getRequestUri()))) {
+            log.error("Internal authentication failed: request signature mismatch");
             request.abortWith(Response.status(UNAUTHORIZED)
                     .type(TEXT_PLAIN_TYPE.toString())
                     .build());
             return;
         }
-        catch (RuntimeException e) {
-            throw new RuntimeException("Authentication error", e);
+
+        if (requestTimestampMillis < 0 || requestTimestampMillis + MAX_REQUEST_AGE_SECONDS * 1000 < currentTimeMillis()) {
+            log.error("Internal authentication failed: request expired at %s", requestTimestampMillis);
+            request.abortWith(Response.status(UNAUTHORIZED)
+                    .type(TEXT_PLAIN_TYPE.toString())
+                    .build());
+            return;
         }
 
         Identity identity = Identity.forUser("<internal>")
-                .withPrincipal(new InternalPrincipal(subject))
+                .withPrincipal(new InternalPrincipal(nodeId))
                 .build();
         setAuthenticatedIdentity(request, identity);
     }
 
+    private String signature(String nodeId, long requestTimestampMillis, URI uri)
+    {
+        return Base64.getEncoder()
+                .encodeToString(hashing.newHasher()
+                        .putUnencodedChars(nodeId)
+                        .putUnencodedChars(uri.toString())
```

Comment:
I'm not sure the full URI is maintained perfectly across the HTTP protocol.  I'd think about using just the path and maybe the query string.